### PR TITLE
refactor: use header info for interchain accounts address generation

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

Removes `UnwrapSDKContext` for interchain accounts. 
Now uses block hash (merkle root of block) instead of dataHash for interchain accounts address generation.

The `ctx.BlockHeader` is deprecated and recommended to now use service based approach. The `HeaderService` provided by cosmos-sdk returns `header.Info` which only contains `appHash`, `hash` (merkle root of block), `height`, `time` and `chainID`.

It is recommended to use `cometservice` when anything else is needed.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

**Review by gjermundgaraba — APPROVED** [review]
Looks good to me! LGTM!

File: modules/apps/27-interchain-accounts/types/account.go [hunk 1]
```diff
@@ -43,15 +43,13 @@ type interchainAccountPretty struct {
...
+func GenerateAddress(headerInfo header.Info, connectionID, portID string) sdk.AccAddress {
 	hostModuleAcc := sdkaddress.Module(ModuleName, []byte(hostAccountsKey))
-	sdkCtx := sdk.UnwrapSDKContext(ctx) // TODO: https://github.com/cosmos/ibc-go/issues/5917
-	header := sdkCtx.BlockHeader()
 
 	buf := []byte(connectionID + portID)
-	buf = append(buf, header.AppHash...)
-	buf = append(buf, header.DataHash...)
+	buf = append(buf, headerInfo.AppHash...)
+	buf = append(buf, headerInfo.Hash...)
```
Comment (AdityaSripal) [inline modules/apps/27-interchain-accounts/types/account.go:52]: I guess we no longer get access to datahash directly? Is there a reason it isn't in header info?

**Review by AdityaSripal — APPROVED** [review]
Looks good in principle. But my understanding is that we can just go back to deterministic address generation because addresses are no longer instantiated as base accounts on receive. So we no longer have to protect against DoS attacks with the new sdk changes

File: modules/apps/27-interchain-accounts/types/account.go (hunk 1 above)
Comment (damiannolan) [reply modules/apps/27-interchain-accounts/types/account.go:52]: That's a question for the sdk team I think.

**Comment by damiannolan** [comment]
> Looks good in principle. But my understanding is that we can just go back to deterministic address generation because addresses are no longer instantiated as base accounts on receive. So we no longer have to protect against DoS attacks with the new sdk changes

Agree! That skipped my mind about the removal of account creation from `SendCoins`. I guess we could removal the block fields entirely from the address generation now. I will merge this PR and that can be discussed later if its something that wants to be done.

**Pull request merged**